File: macos/Turkify/Turkify/AppSettings.swift
import Foundation

/// macOS uygulamasının ayarları — **native saklanır** (`UserDefaults`), config.json
/// **kullanılmaz**. (config.json yalnızca Linux servisi ve CLI içindir.)
///
/// Ayarlar motora `turkify serve`'e **CLI bayrakları** olarak geçirilir; bayraklar
/// önceliklidir (CLI > env > config), bu yüzden bir config.json olsa bile GUI
/// ayarları kazanır. Ayar değişince motor yeni bayraklarla yeniden başlatılır.
struct AppSettings {
    var model: String
    var useLLM: Bool
    var useMorphology: Bool
    /// Cümle sonu noktalamadan (.!?…) sonra küçük harfi büyütür. Varsayılan kapalı.
    var capitalizeSentences: Bool
    /// `capitalizeSentences` açıkken metnin (seçimin) ilk harfini de büyütür.
    /// Bağımlı ayar: üst ayar kapalıysa motorda etkisizdir. Varsayılan kapalı.
    var capitalizeFirst: Bool
    var timeout: Int  // saniye (tam sayı)
    var baseURL: String
    var apiKey: String
    var assistantPrefill: String
    /// /chat/completions gövdesine eklenecek ham JSON (boş = gönderilmez).
    var llmOptions: String
    var hotkeyMods: [String]       // düzeltme kısayolu
    var hotkeyKey: String
    var cancelHotkeyMods: [String] // işlem iptali kısayolu
    var cancelHotkeyKey: String
    /// Yalnızca UI tercihi (motora gönderilmez): uygulama çalışırken Dock'ta
    /// simge görünsün mü? Kapalıysa menü-bar-only (accessory).
    var showInDock: Bool
    /// UI tercihi: model seçimi otomatik mi (combobox'tan, model/url readonly) yoksa
    /// manuel mi (combobox kapalı, model/url elle yazılır)?
    var autoModelSelection: Bool

    static let fallback = AppSettings(
        model: "",
        useLLM: false,
        useMorphology: true,
        capitalizeSentences: false,
        capitalizeFirst: false,
        timeout: 60,
        baseURL: "http://localhost:11434/v1",
        apiKey: "",
        assistantPrefill: "",
        llmOptions: "",
        hotkeyMods: ["ctrl", "opt", "cmd"],
        hotkeyKey: "a",
        cancelHotkeyMods: ["ctrl", "opt", "cmd"],
        cancelHotkeyKey: "q",
        showInDock: false,
        autoModelSelection: true
    )

    static func load() -> AppSettings {
        let d = UserDefaults.standard
        return AppSettings(
            model: d.string(forKey: "model") ?? "",
            useLLM: d.object(forKey: "useLLM") as? Bool ?? fallback.useLLM,
            useMorphology: d.object(forKey: "useMorphology") as? Bool ?? fallback.useMorphology,
            capitalizeSentences: d.object(forKey: "capitalizeSentences") as? Bool ?? fallback.capitalizeSentences,
            capitalizeFirst: d.object(forKey: "capitalizeFirst") as? Bool ?? fallback.capitalizeFirst,
            timeout: (d.object(forKey: "timeout") as? NSNumber)?.intValue ?? fallback.timeout,
            baseURL: d.string(forKey: "baseURL") ?? fallback.baseURL,
            apiKey: d.string(forKey: "apiKey") ?? "",
            assistantPrefill: d.string(forKey: "assistantPrefill") ?? "",
            llmOptions: d.string(forKey: "llmOptions") ?? "",
            hotkeyMods: d.stringArray(forKey: "hotkeyMods") ?? fallback.hotkeyMods,
            hotkeyKey: d.string(forKey: "hotkeyKey") ?? fallback.hotkeyKey,
            cancelHotkeyMods: d.stringArray(forKey: "cancelHotkeyMods") ?? fallback.cancelHotkeyMods,
            cancelHotkeyKey: d.string(forKey: "cancelHotkeyKey") ?? fallback.cancelHotkeyKey,
            showInDock: d.object(forKey: "showInDock") as? Bool ?? fallback.showInDock,
            autoModelSelection: d.object(forKey: "autoModelSelection") as? Bool ?? fallback.autoModelSelection
        )
    }

    func save() {
        let d = UserDefaults.standard
        d.set(model, forKey: "model")
        d.set(useLLM, forKey: "useLLM")
        d.set(useMorphology, forKey: "useMorphology")
        d.set(capitalizeSentences, forKey: "capitalizeSentences")
        d.set(capitalizeFirst, forKey: "capitalizeFirst")
        d.set(timeout, forKey: "timeout")
        d.set(baseURL, forKey: "baseURL")
        d.set(apiKey, forKey: "apiKey")
        d.set(assistantPrefill, forKey: "assistantPrefill")
        d.set(llmOptions, forKey: "llmOptions")
        d.set(hotkeyMods, forKey: "hotkeyMods")
        d.set(hotkeyKey, forKey: "hotkeyKey")
        d.set(cancelHotkeyMods, forKey: "cancelHotkeyMods")
        d.set(cancelHotkeyKey, forKey: "cancelHotkeyKey")
        d.set(showInDock, forKey: "showInDock")
        d.set(autoModelSelection, forKey: "autoModelSelection")
    }

    /// Motoru bu ayarlarla başlatacak `serve` argümanları (CLI bayrakları).
    func serveArguments() -> [String] {
        // --verbose: motorun Tier 2/3 karar günlüğünü stderr'e yazdırır; bu çıktı
        // EngineClient tarafından yakalanıp Log sekmesine (motor kaynağı) düşer.
        // stdout (JSON protokolü) temiz kalır.
        var args = ["serve", "--stdio", "--verbose"]
        if !model.isEmpty { args += ["--model", model] }
        args += [useLLM ? "--llm" : "--no-llm"]
        args += [useMorphology ? "--morphology" : "--no-morphology"]
        args += [capitalizeSentences ? "--capitalize-sentences" : "--no-capitalize-sentences"]
        args += [capitalizeFirst ? "--capitalize-first" : "--no-capitalize-first"]
        args += ["--timeout", String(timeout)]
        if !baseURL.isEmpty { args += ["--base-url", baseURL] }
        if !apiKey.isEmpty { args += ["--api-key", apiKey] }
        if !assistantPrefill.isEmpty { args += ["--assistant-prefill", assistantPrefill] }
        // Yalnızca geçerli JSON ise gönder; geçersizse motor hiç başlamazdı.
        let trimmedOptions = llmOptions.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmedOptions.isEmpty, AppSettings.isValidJSON(trimmedOptions) {
            args += ["--llm-options", trimmedOptions]
        }
        return args
    }

    /// Boş ya da geçerli JSON mu? (UI'da uyarı + serveArguments'ta filtre için.)
    static func isValidJSON(_ text: String) -> Bool {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return true }
        guard let data = trimmed.data(using: .utf8) else { return false }
        return (try? JSONSerialization.jsonObject(with: data)) != nil
    }

    var hotkeyDescription: String {
        Self.describe(mods: hotkeyMods, key: hotkeyKey)
    }

    var cancelHotkeyDescription: String {
        Self.describe(mods: cancelHotkeyMods, key: cancelHotkeyKey)
    }

    private static func describe(mods: [String], key: String) -> String {
        mods.map { $0.capitalized }.joined(separator: "+") + "+" + key.uppercased()
    }

    /// Motoru çalıştıracak komut. `serve` argümanları ayrı eklenir.
    ///
    /// Öncelik:
    ///  1. **Geliştirme:** Xcode scheme'inde `TURKIFY_PYTHON` env'i (venv python + `-m turkify`).
    ///  2. **Release:** `.app` içine gömülü donmuş motor
    ///     (`Contents/Resources/turkify-engine/turkify-engine`; bkz. ADR 0009 / packaging).
    ///  3. **Son çare:** sistem `python3 -m turkify` (yalnızca geliştirme fallback'i).
    static func engineExecutable() -> (executable: String, prefixArgs: [String]) {
        if let python = ProcessInfo.processInfo.environment["TURKIFY_PYTHON"], !python.isEmpty {
            return (python, ["-m", "turkify"])
        }
        if let embedded = Bundle.main.url(
            forResource: "turkify-engine", withExtension: nil, subdirectory: "turkify-engine"
        ) {
            return (embedded.path, [])
        }
        return ("/usr/bin/env", ["python3", "-m", "turkify"])
    }
}

File: windows/Turkify/AppSettings.cs
using System.IO;
using System.Text.Json;
using Microsoft.Win32;

namespace Turkify;

/// Windows uygulamasının ayarları — **native saklanır** (Registry
/// <c>HKCU\Software\Turkify</c>), config.json **kullanılmaz** (yalnızca Linux/CLI
/// içindir; bkz. ADR 0007). Ayarlar motora <c>turkify serve</c>'e **CLI bayrakları**
/// olarak geçirilir; bayraklar önceliklidir (CLI > env > config), bu yüzden bir
/// config.json olsa bile GUI ayarları kazanır. Ayar değişince motor yeni bayraklarla
/// yeniden başlatılır. (macOS AppSettings.swift'in karşılığı.)
public sealed class AppSettings
{
    public string Model { get; set; } = "";
    public bool UseLLM { get; set; }
    public bool UseMorphology { get; set; } = true;
    public int Timeout { get; set; } = 60; // saniye
    public string BaseURL { get; set; } = "http://localhost:11434/v1";
    public string ApiKey { get; set; } = "";
    public string AssistantPrefill { get; set; } = "";

    /// /chat/completions gövdesine eklenecek ham JSON (boş = gönderilmez).
    public string LlmOptions { get; set; } = "";

    // Varsayılan: Hyper (Ctrl+Alt+Win) — macOS'taki Hyper (Ctrl+Opt+Cmd) ile aynı,
    // cmd→Win. Düz Win-kombinasyonlarının OS çakışmasını azaltır.
    public string[] HotkeyMods { get; set; } = ["ctrl", "alt", "win"]; // düzeltme kısayolu
    public string HotkeyKey { get; set; } = "a";
    public string[] CancelHotkeyMods { get; set; } = ["ctrl", "alt", "win"]; // işlem iptali
    public string CancelHotkeyKey { get; set; } = "q";

    /// UI tercihi: model seçimi otomatik mi (combobox'tan, model/url salt-okunur)
    /// yoksa manuel mi (elle yazılır)?
    public bool AutoModelSelection { get; set; } = true;

    private const string RegistryKeyPath = @"Software\Turkify";

    public static AppSettings Load()
    {
        using RegistryKey? key = Registry.CurrentUser.OpenSubKey(RegistryKeyPath);
        if (key is null)
        {
            return new AppSettings();
        }

        var fallback = new AppSettings();
        return new AppSettings
        {
            Model = GetString(key, "Model", fallback.Model),
            UseLLM = GetBool(key, "UseLLM", fallback.UseLLM),
            UseMorphology = GetBool(key, "UseMorphology", fallback.UseMorphology),
            Timeout = GetInt(key, "Timeout", fallback.Timeout),
            BaseURL = GetString(key, "BaseURL", fallback.BaseURL),
            ApiKey = GetString(key, "ApiKey", fallback.ApiKey),
            AssistantPrefill = GetString(key, "AssistantPrefill", fallback.AssistantPrefill),
            LlmOptions = GetString(key, "LlmOptions", fallback.LlmOptions),
            HotkeyMods = GetMods(key, "HotkeyMods", fallback.HotkeyMods),
            HotkeyKey = GetString(key, "HotkeyKey", fallback.HotkeyKey),
            CancelHotkeyMods = GetMods(key, "CancelHotkeyMods", fallback.CancelHotkeyMods),
            CancelHotkeyKey = GetString(key, "CancelHotkeyKey", fallback.CancelHotkeyKey),
            AutoModelSelection = GetBool(key, "AutoModelSelection", fallback.AutoModelSelection),
        };
    }

    public void Save()
    {
        using RegistryKey key = Registry.CurrentUser.CreateSubKey(RegistryKeyPath);
        key.SetValue("Model", Model);
        key.SetValue("UseLLM", UseLLM ? "1" : "0");
        key.SetValue("UseMorphology", UseMorphology ? "1" : "0");
        key.SetValue("Timeout", Timeout.ToString());
        key.SetValue("BaseURL", BaseURL);
        key.SetValue("ApiKey", ApiKey);
        key.SetValue("AssistantPrefill", AssistantPrefill);
        key.SetValue("LlmOptions", LlmOptions);
        key.SetValue("HotkeyMods", string.Join(' ', HotkeyMods));
        key.SetValue("HotkeyKey", HotkeyKey);
        key.SetValue("CancelHotkeyMods", string.Join(' ', CancelHotkeyMods));
        key.SetValue("CancelHotkeyKey", CancelHotkeyKey);
        key.SetValue("AutoModelSelection", AutoModelSelection ? "1" : "0");
    }

    /// Motoru bu ayarlarla başlatacak <c>serve</c> argümanları (CLI bayrakları).
    /// macOS serveArguments() ile aynı bayrak kümesi.
    public IReadOnlyList<string> ServeArguments()
    {
        // --verbose: motorun Tier 2/3 karar günlüğünü stderr'e yazdırır; bu çıktı
        // EngineClient tarafından yakalanıp Log'a (motor kaynağı) düşer. stdout
        // (JSON protokolü) temiz kalır.
        var args = new List<string> { "serve", "--stdio", "--verbose" };
        if (!string.IsNullOrEmpty(Model))
        {
            args.Add("--model");
            args.Add(Model);
        }

        args.Add(UseLLM ? "--llm" : "--no-llm");
        args.Add(UseMorphology ? "--morphology" : "--no-morphology");
        args.Add("--timeout");
        args.Add(Timeout.ToString());

        if (!string.IsNullOrEmpty(BaseURL))
        {
            args.Add("--base-url");
            args.Add(BaseURL);
        }

        if (!string.IsNullOrEmpty(ApiKey))
        {
            args.Add("--api-key");
            args.Add(ApiKey);
        }

        if (!string.IsNullOrEmpty(AssistantPrefill))
        {
            args.Add("--assistant-prefill");
            args.Add(AssistantPrefill);
        }

        // Yalnızca geçerli JSON ise gönder; geçersizse motor hiç başlamazdı.
        string trimmedOptions = LlmOptions.Trim();
        if (trimmedOptions.Length > 0 && IsValidJson(trimmedOptions))
        {
            args.Add("--llm-options");
            args.Add(trimmedOptions);
        }

        return args;
    }

    /// Boş ya da geçerli JSON mu? (UI'da uyarı + serveArguments'ta filtre için.)
    public static bool IsValidJson(string text)
    {
        string trimmed = text.Trim();
        if (trimmed.Length == 0)
        {
            return true;
        }

        try
        {
            using var _ = JsonDocument.Parse(trimmed);
            return true;
        }
        catch (JsonException)
        {
            return false;
        }
    }

    public string HotkeyDescription => Describe(HotkeyMods, HotkeyKey);

    public string CancelHotkeyDescription => Describe(CancelHotkeyMods, CancelHotkeyKey);

    private static string Describe(string[] mods, string key) =>
        string.Join('+', mods.Select(Capitalize).Append(key.ToUpperInvariant()));

    private static string Capitalize(string s) =>
        s.Length == 0 ? s : char.ToUpperInvariant(s[0]) + s[1..];

    /// Motoru çalıştıracak komut. <c>serve</c> argümanları ayrı eklenir.
    ///
    /// Öncelik:
    ///  1. **Geliştirme:** <c>TURKIFY_PYTHON</c> env'i (venv python + <c>-m turkify</c>).
    ///  2. **Release:** uygulama klasörüne gömülü donmuş motor
    ///     (<c>turkify-engine\turkify-engine.exe</c>; bkz. ADR 0009 / packaging).
    ///  3. **Son çare:** sistem <c>python -m turkify</c> (yalnızca geliştirme fallback'i).
    public static (string Executable, IReadOnlyList<string> PrefixArgs) EngineExecutable()
    {
        string? python = Environment.GetEnvironmentVariable("TURKIFY_PYTHON");
        if (!string.IsNullOrEmpty(python))
        {
            return (python, ["-m", "turkify"]);
        }

        string baseDir = AppContext.BaseDirectory;
        string embedded = Path.Combine(baseDir, "turkify-engine", "turkify-engine.exe");
        if (File.Exists(embedded))
        {
            return (embedded, []);
        }

        return ("python", ["-m", "turkify"]);
    }

    private static string GetString(RegistryKey key, string name, string fallback) =>
        key.GetValue(name) as string ?? fallback;

    private static bool GetBool(RegistryKey key, string name, bool fallback) =>
        key.GetValue(name) is string s ? s == "1" : fallback;

    private static int GetInt(RegistryKey key, string name, int fallback) =>
        key.GetValue(name) is string s && int.TryParse(s, out int v) ? v : fallback;

    private static string[] GetMods(RegistryKey key, string name, string[] fallback) =>
        key.GetValue(name) is string s
            ? s.Split(' ', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            : fallback;
}
